main.js

```diff
@@ -1,186 +1,256 @@
+		},
+
+		// is an engine available in a array of symbolGroups
+		getSymbolsForEngine(symbolGroups, symbol, engineId) {
+			if(!symbolGroups.length) return false;
+			var filteredGroups = symbolGroups.filter(function(symbols) {
+				if(!symbols || !symbols[symbol]) return false;
+				var engines =	 symbols[symbol].engines;
+				var fns =	 symbols[symbol].fns;
+				if (engines) {
+					return engines[engineId] ? true : false
+				}
+				if (fns) {
+					return fns[engineId] ? true : false
+				}
+
+				return false;
+			})
+			if(!filteredGroups.length) return false
+			return filteredGroups[0]
+		},
+
+		// param:
+		// - userQuery: string `!m new york city`
+		// return:
+		// - url: string to be openned by the browser
+		decodeUserRequest(userRequest) {
+			if(!userRequest) { return false; }
 
-App.init();
+			var requestTerms = userRequest.split(' '),
+					requestSymbolGroup = requestTerms[0],
+					requestSymbol = this.checkForSymbol(requestSymbolGroup),
+					requestEngineId = requestSymbolGroup.slice(1),
+					allSymbolGroups = [this.getUserSymbols(), this.symbols];
+
+			// if there is no symbol, the whole userRequest is the query
+			if (!requestSymbol) {
+				return this.buildResultUrl(userRequest);
+			}
+
+			// is the engine referenced in the	userSymbols or symbols
+			// let selectedSymbols = this.getRequestedSymbols(allSymbolGroups, requestSymbol, requestEngineId);
+			var selectedSymbols = this.getSymbolsForEngine(allSymbolGroups, requestSymbol, requestEngineId);
+
+			// if there are no selectedSymbols, we don't know the engine
+			if (!selectedSymbols) {
+				return this.buildResultUrl(userRequest);
+			}
+
+			// if we know the symbol and engine,
+			// the actual query is everything but the request's engine group (the first group)
+			var userRequestNoSymbol = requestTerms.splice(1, requestTerms.length).join(' ');
+			return this.buildResultUrl(userRequestNoSymbol, selectedSymbols, requestSymbol, requestEngineId);
+		},
+
+		// check whether URL starts with a scheme
+		checkUrl(url) {
+			return url.startsWith("//") || url.includes("://");
+		},
+
+		openUrl(url) {
+			// replace history state
+			// so after transition, clicking the back button does not hit find/?q=search
+			// that would transition again to a search result
+			if(!url) return
+			if (!this.checkUrl(url)) url = "//" + url;
+			location.replace(url);
+		},
+
+		// takes a string, request query of a user
+		find(request) {
+			if(!request) return false;
+			return this.openUrl(this.decodeUserRequest(request));
+		},
+
+		init() {
+			var url = new URL(window.location.href);
+			var request = url.searchParams.get('q');
+			if(!request) return;
+			this.find(request);
+		},
+
+		// get the user symbols from local storage
+		// or returns an empty new set of symbols
+		getUserSymbols() {
+			var storageSymbols;
+			try {
+				storageSymbols = JSON.parse(
+					localStorage.getItem(localStorageKey)
+				);
+			} catch(e) {
+				if(e.name === 'SyntaxError') {
```

Comment:
Why does this only treat syntax errors?

An immediate problem with this is that if another error occurs, line 234 (`if(!storageSymbols) ...`) will error because the variable is not definied.

Reply:
The idea was that `SyntaxError` will show if parsing the json did not work.
This would take care of the case the data is broken in the store.
Maybe we can have it for all errors instead.